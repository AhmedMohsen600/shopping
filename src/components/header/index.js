import { Outlet, Link, useLocation } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { useState } from 'react';
import { fetchProducts } from '../../features/products/productsSlice';
import {
  Container,
  Nav,
  BackArrow,
  Categories,
  Category,
} from './styles/header';
import data from '../../fixtures/categories.json';

export default function Header() {
  const dispatch = useDispatch();
  const { pathname } = useLocation();
  const [current, setCurrent] = useState(
    JSON.parse(localStorage.getItem('category')) || 'electronics'
  );
  const productLength = useSelector((state) => state.products.data.length);

  return (
    <>
      <Container>
        <Nav>
          <BackArrow to="/">
            {pathname === '/' ? `${current}` : 'Back'}
          </BackArrow>
          <Categories>
            {data.map((categ) => (
              <Link className="link" to="/" key={categ}>
                <Category
                  onClick={() => {
                    localStorage.setItem('category', JSON.stringify(categ));
                    if (categ === current) return;
                    setCurrent(categ);
                    dispatch(fetchProducts(categ));
                  }}
                  active={categ === current}
                >
                  {categ}
                </Category>
              </Link>
            ))}
          </Categories>
        </Nav>
      </Container>
      <Outlet />
    </>
  );
}
