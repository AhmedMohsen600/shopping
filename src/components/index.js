export { default as Product } from './product';

export { default as Header } from './header';

export { default as ProductDetails } from './product-details';
