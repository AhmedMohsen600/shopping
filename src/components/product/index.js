export default function Product() {
  return <></>;
}
