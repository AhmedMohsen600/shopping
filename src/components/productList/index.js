import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchProducts } from '../../features/products/productsSlice';
import { Container, Inner } from './styles/product-list';
import Product from '../product';

export default function ProductList() {
  const dispatch = useDispatch();
  const products = useSelector((state) => state.products.data);
  const isLoading = useSelector((state) => state.products.loading);

  useEffect(() => {
    if (products.length) return;
    dispatch(
      fetchProducts(
        JSON.parse(localStorage.getItem('category')) || 'electronics'
      )
    );
  }, [dispatch, products.length]);

  return isLoading ? (
    <div style={{ fontSize: '40px', textAlign: 'center' }}>Loading</div>
  ) : (
    <Container>
      <Inner>
        {products.map(({ image, title, description, category, id, price }) => (
          <Product
            key={id}
            src={image}
            title={title}
            description={description}
            category={category}
            price={price}
            id={id}
          />
        ))}
      </Inner>
    </Container>
  );
}
