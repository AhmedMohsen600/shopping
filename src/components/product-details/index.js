import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import {
  Container,
  Inner,
  Image,
  ImageHolder,
  DetailsHolder,
  ProductDesc,
  ProducTitle,
} from './styles/product-details';
import { Price } from '../product/styles/product';
import { fetchProduct } from '../../features/product-details/productDetailstSlice';

export default function ProductDetails() {
  const { pathname } = useLocation();
  const dispatch = useDispatch();
  const product = useSelector((state) => state.product.data);
  const isLoading = useSelector((state) => state.product.loading);
  const { image, title, description, price, category, rating } = product;
  useEffect(() => {
    dispatch(fetchProduct(pathname.split('/')[3]));
  }, [dispatch, pathname]);

  return isLoading ? (
    <div style={{ fontSize: '40px', textAlign: 'center' }}>Loading</div>
  ) : (
    <Container>
      <Inner>
        <ImageHolder>
          <Image src={image} />
        </ImageHolder>
        <DetailsHolder>
          <ProducTitle>{product ? title : 'product name'}</ProducTitle>
          <ProductDesc>{product ? description : 'Description'}</ProductDesc>
          <Price>Price: {product ? `$${price}` : 'Price'}</Price>
          <Price>Category: {product ? category : 'Category'}</Price>
          <Price>Rate: {rating ? rating.rate : 'Rate'}</Price>
          <Price>Count: {rating ? rating.count : 'Count'}</Price>
        </DetailsHolder>
      </Inner>
    </Container>
  );
}
