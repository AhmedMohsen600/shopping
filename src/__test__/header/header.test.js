import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Header } from '../../components';

jest.mock('react-redux');
describe('Test component', () => {
  test('Test Header components', () => {
    const tree = render(
      <BrowserRouter>
        <Header />
      </BrowserRouter>
    );
    expect(tree).toMatchSnapshot();
  });
});
