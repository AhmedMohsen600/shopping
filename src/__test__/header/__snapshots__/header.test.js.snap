// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component Test Header components 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header__Container-sc-nre3gu-0 fFDyFw"
      >
        <nav
          class="header__Nav-sc-nre3gu-1 LAEPW"
        >
          <a
            class="header__BackArrow-sc-nre3gu-3 dbGXyx"
            href="/"
          >
            electronics (undefined)
          </a>
          <ul
            class="header__Categories-sc-nre3gu-4 gphpXY"
          >
            <a
              class="link"
              href="/"
            >
              <li
                class="header__Category-sc-nre3gu-5 igYrWh"
              >
                electronics
              </li>
            </a>
            <a
              class="link"
              href="/"
            >
              <li
                class="header__Category-sc-nre3gu-5 bNmxvb"
              >
                jewelery
              </li>
            </a>
            <a
              class="link"
              href="/"
            >
              <li
                class="header__Category-sc-nre3gu-5 bNmxvb"
              >
                men's clothing
              </li>
            </a>
            <a
              class="link"
              href="/"
            >
              <li
                class="header__Category-sc-nre3gu-5 bNmxvb"
              >
                women's clothing
              </li>
            </a>
          </ul>
        </nav>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="header__Container-sc-nre3gu-0 fFDyFw"
    >
      <nav
        class="header__Nav-sc-nre3gu-1 LAEPW"
      >
        <a
          class="header__BackArrow-sc-nre3gu-3 dbGXyx"
          href="/"
        >
          electronics (undefined)
        </a>
        <ul
          class="header__Categories-sc-nre3gu-4 gphpXY"
        >
          <a
            class="link"
            href="/"
          >
            <li
              class="header__Category-sc-nre3gu-5 igYrWh"
            >
              electronics
            </li>
          </a>
          <a
            class="link"
            href="/"
          >
            <li
              class="header__Category-sc-nre3gu-5 bNmxvb"
            >
              jewelery
            </li>
          </a>
          <a
            class="link"
            href="/"
          >
            <li
              class="header__Category-sc-nre3gu-5 bNmxvb"
            >
              men's clothing
            </li>
          </a>
          <a
            class="link"
            href="/"
          >
            <li
              class="header__Category-sc-nre3gu-5 bNmxvb"
            >
              women's clothing
            </li>
          </a>
        </ul>
      </nav>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
